snuba/datasets/configuration/entity_builder.py

```diff
@@ -0,0 +1,135 @@
+from __future__ import annotations
+
+import logging
+from typing import Any, Mapping, Sequence, Type
+
+from snuba.clickhouse.translators.snuba.allowed import (
+    FunctionCallMapper,
+    SubscriptableReferenceMapper,
+)
+from snuba.clickhouse.translators.snuba.mappers import (
+    FunctionNameMapper,
+    SubscriptableMapper,
+)
+from snuba.clickhouse.translators.snuba.mapping import TranslationMappers
+from snuba.datasets.configuration.json_schema import V1_ENTITY_SCHEMA
+from snuba.datasets.configuration.loader import load_configuration_data
+from snuba.datasets.configuration.utils import parse_columns
+from snuba.datasets.entity import Entity
+from snuba.datasets.pluggable_entity import PluggableEntity
+from snuba.datasets.storages import StorageKey
+from snuba.datasets.storages.factory import get_storage, get_writable_storage
+from snuba.query.processors import QueryProcessor
+from snuba.query.validation.validators import (
+    EntityRequiredColumnValidator,
+    QueryValidator,
+)
+
+_QP_MAPPING: dict[str, Type[QueryProcessor]] = {}
+
+_FUNCTION_MAPPER_MAPPING: Mapping[str, Type[FunctionCallMapper]] = {
+    "simple_func": FunctionNameMapper
+}
+
+_SUB_MAPPER_MAPPING: Mapping[str, Type[SubscriptableReferenceMapper]] = {
+    "subscriptable": SubscriptableMapper
+}
+
+_VALIDATOR_MAPPING: Mapping[str, Type[QueryValidator]] = {
+    "entity_required": EntityRequiredColumnValidator
+}
+
+logger = logging.getLogger("snuba.entity_builder")
+
+
+def _initialize_mappings() -> None:
```

Comment:
this pattern can cause threading issues. instead, have the function return the constructed object.

```python3
_QP_MAPPING = None

def _qp_mapping() -> dict[str, Type[QueryProcessor]]:
    global _QP_MAPPING
    if not _QP_MAPPING:
        # initialize here
    return _QP_MAPPING
```

This way if there's a race condition, worst case is you'll have two threads construct and return the same object

Reply:
I actually don't see the functional difference between my approach and yours, assuming that dict.update is atomic.

In my approach:
```
t=0, Thread T1 sees _QP_MAPPING with length 0
t=1, Thread T2 sees _QP_MAPPING with length 0
t=2, Thread T2 finishes importing
t=3, Thread T1 finishes importing
t=4, Thread T2 calls _QP_MAPPING.update(constants)
t=5, Thread T1 calls _QP_MAPPING.update(constants)
```
extra update, no data inconsistency

Your approach:
```
t=0, Thread T1 sees _QP_MAPPING != None
t=1, Thread T2 sees _QP_MAPPING != None
t=2, Thread T2 finishes importing
t=3, Thread T1 finishes importing
t=4, Thread T2 sets _QP_MAPPING = constants
t=5, Thread T1 sets _QP_MAPPING = constants
t=6 or 7, Threads T1 and T2 should return _QP_MAPPING with the same contents
```

I think the issue with my code before is that within that `# initialize here` section of yours I was re-checking the original condition and the value changed and I mishandled it.

Also, we hope to remove this code (or at least move out of this file into the metaclass/QueryProcessor definition) when we go with the registered class pattern

Reply:
I don't think that `_QP_MAPPING.update` is atomic (I could be wrong). So you could see `_QP_MAPPING` be partially updated but not fully.

Reply:
There are other benefits to this approach as well. In a test, you can mock the return value of the function to return what you want in that instance of the test. It's also a good practice in general to hide accesses of a global variable behind a function.

Reply:
i have absolutely no stake in this but this particular thread caught my eye. `dict.update` is holding the GIL continuously (= "is atomic") for as long as the dictionary keys are primitive types and their `__hash__` is written in C. This is the case for `str`, so this case is fine.

_If_ you happen to update with a dictionary key whose `__hash__` is written in _Python_, you will release the GIL and could probably maybe observe inconsistent state. I think in this case it would mean that one thread will observe its dictionary being mutated, updated with values that are equal with the previous ones, but have a different `id()`.

A variable assignment is guaranteed to be atomic no matter what the variable type is.